File: 0540 Single Element in a Sorted Array/single_element_in_a_sorted_array.cpp
/*
BY RUSHABH GALA
rushabhvg
GitHub :
https://github.com/rushabhvg
StackOverflow :
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/

class Solution {
    int start = 0;
    int last;
    int mid;
public:
    void binSrch(vector<int>& nums) {
        while(start<last) {
            mid = start + (last-start)/2;
            if (nums[mid]==nums[mid^1]) {
                start = mid+1;
            } else {
                last = mid;
            }
        }
    }
    int singleNonDuplicate(vector<int>& nums) {
        last = nums.size();
        if(last==1) return nums[0];
        binSrch(nums);
        return nums[start];
    }
};

/*
BY RUSHABH GALA
rushabhvg
GitHub :
https://github.com/rushabhvg
StackOverflow :
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/

File: 540-single-element-in-a-sorted-array/single-element-in-a-sorted-array.java
/*
By Rushabh Gala
rushabhvg
GitHub:
https://github.com/rushabhvg
StackOverflow:
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/

class Solution {
    public int singleNonDuplicate(int[] nums) {
        int n = nums.length;
        if(n==1) return nums[0];
        int st = 0;
        int end = n-1;
        int mid = 0;
        while(st<end) {
            mid = st + (end-st)/2;
            if(mid%2==1) {
                mid--;
            }
            if( nums[mid]!=nums[mid+1] ) {
                end = mid;
            } else {
                st = mid+2;
            }
        }
        return nums[st];
    }
}

/*
By Rushabh Gala
rushabhvg
GitHub:
https://github.com/rushabhvg
StackOverflow:
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/
